## ECCB 2018 Workshop - Exploring Programmatic Access to Protein Sequence, Function and Structure Information

### Summary

Integrating publicly available biological data from multiple data sources (including your own) can be critical to analyse data and infer patterns that may not otherwise be obvious. In this workshop you will learn how to discover and integrate biological information as well as visualisation components from protein resources at the European Bioinformatics Institute (EMBL-EBI), with a focus on UniProt and PDBe. 

Learn about accessing data such as gene-protein relationships, protein sequences, protein structures and function. Explore web services, like the [UniProt API](https://www.ebi.ac.uk/proteins/api/doc/) or the [PDBe API]( http://www.ebi.ac.uk/pdbe/api/doc/) providing programmatic access to data as well as freely available visualisation components, such as [PDBe’s LiteMol viewer](https://www.ebi.ac.uk/pdbe/pdb-component-library/doc.html#a_LiteMol) and [UniProt’s ProtVista](http://ebi-uniprot.github.io/ProtVista/) that you can re-use to visualise this data as well as your own.

### Workshop schedule

#### Morning Block 01
09:00 - 09.20 Introductions and context

09:20 - 09:40 UniProt intro

09:40 - 10:00 PDBe intro

10:00 - 10:30 Guided hands-on exploration of web pages (3x10min case studies)

##### Coffee break

#### Morning Block 02
11:00 - 11:20 Introduction to UniProt API

11:20 - 11:40 Introduction to PDBe API

11:40 - 12:30 Guided hands-on exploration (Jupyter notebooks)

##### Lunch break

#### Afternoon Block 01
13:30 - 13:50 UniProt open access visualisation components

13:50 - 14:10 PDBe web component library

14:10 - 15:00 Guided hands-on exploration (HTML templates)

##### Coffee break

#### Afternoon Block 02
15:30 - 17:00	Group workshop (Data analysis problems)

### Tutors:
#### Andrew Nightingale, EMBL-EBI, Wellcome Genome Campus Cambridge, UK
![Andrew Nightingale](https://avatars2.githubusercontent.com/u/38358380?s=100&v=4)
I have received a BSc honours degree in Biochemistry from The University of Bath in 1997, a MSc in Computational Biology from the University of York in 1998 and a Ph.D. in Biochemistry/Bioinformatics in 2002 from The University of Leeds; where I worked in the Dr Howard Parish and Professor David Westhead groups. I then joined ProSpect Pharma Inc as a computational biologist and research scientist specialising in the development of new techniques for structure based drug design by NMR in collaboration with the Professor Steven Homans group. In 2012, I joined the UniProt team at EMBL-EBI as a data scientist responsible for the identification, analysis and integration of data resources related to genomic annotation, variation and active biomolecules to expand the UniProt Knowledgebase (UniProtKB) for clinical research and the understanding of the molecular mechanisms of diseases. This has led to the development of new data resources and tools such as the RESTful Proteins API, UniProt annotations available for integration into genome browsers and a genomic coordinates service for mapping protein expressing genes to their expressed proteins and annotations. I have composed and presented workshops on behalf of EMBL-EBI on the access and use of UniProt variation data; including a two day workshop at The Colombian Center for Bioinformatics and Computational Biology (BIOS) targeted at pre-doctoral and post-doctoral fellows interested in analysing the functional consequences of variants. I have also written and presented an EMBL-EBI webinar on programmatic access to UniProt data via its RESTful Proteins API.

#### Mihaly Varadi, EMBL-EBI, Wellcome Genome Campus, Cambridge, UK
![Mihaly Varadi](https://avatars3.githubusercontent.com/u/24314338?s=100)

Background: I received my MSc in Applied Zoology from the Veterinary University of Budapest in 2011, and my PhD in Bioscience Engineering/Computational Biology from the Vrije Universiteit Brussel in 2015. In the same year I have joined the Protein Data Bank in Europe at the European Bioinformatics Institute (EMBL-EBI) as scientific programmer. Since then I have been working on the back-end of the PDB as well as on providing programmatic access to the data via APIs, and displaying information on our front-ends using web components. I work with Python/Django, Java/Neo4J and TypeScript/Angular&Polymer.

### Course Material


### Group Workshop topics
