## ECCB 2018 Workshop - Exploring programmatic access to Protein sequence, function and structure information

### Summary

Integrating publicly available biological data from multiple data sources (including your own) can be critical to analyse data and infer patterns that may not otherwise be obvious. Learn how to discover and integrate biological information as well as visualisation components from protein resources at the European Bioinformatics Institute (EMBL-EBI), with a focus on UniProt and PDBe. Learn about accessing data such as gene-protein relationships, protein sequences, protein structures and function. Explore newly created services (such as https://www.ebi.ac.uk/proteins/api/doc/ http://www.ebi.ac.uk/pdbe/api/doc/) providing programmatic access to this data as well as freely available visualisation components, such as PDBe’s LiteMol viewer (https://www.ebi.ac.uk/pdbe/pdb-component-library/doc.html#a_LiteMol) and UniProt’s ProtVista (http://ebi-uniprot.github.io/ProtVista/) that you can re-use to visualise this data as well as your own.


### Tutors:
#### Andrew Nightingale, EMBL-EBI, Wellcome Genome Campus Cambridge, UK

#### Mihaly Varadi, EMBL-EBI, Wellcome Genome Campus, Cambridge, UK
[<img src="https://www.ebi.ac.uk/sites/ebi.ac.uk/files/styles/medium/public/person/image/varadi_mihaly_web_2015.jpg">](https://www.ebi.ac.uk/about/people/mihaly-varadi)
Background: I received my MSc in Applied Zoology from the Veterinary University of Budapest in 2011, and my PhD in Bioscience Engineering/Computational Biology from the Vrije Universiteit Brussel in 2015. In the same year I have joined the Protein Data Bank in Europe at the European Bioinformatics Institute (EMBL-EBI) as scientific programmer. Since then I have been working on the back-end of the PDB as well as on providing programmatic access to the data via APIs, and displaying information on our front-ends using web components. I work with Python/Django, Java/Neo4J and TypeScript/Angular&Polymer.

### Course Material


### Group Workshop topics
