{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example Python Script to GET genomic coordinates and processing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GET request and return of JSON"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This example shows a basic way to handle paginated results from the API. Not all Proteins API end-points return paginated data.\n",
    "\n",
    "One __important__ method is _checkResponse_, as you need to check that the service is up and returning usable data. Response code __200__ means that the service is up and data has been returned. Other response codes can be used to check for different responses.\n",
    "\n",
    "The _jsonRequest_ method is allowed to have up to 10 tries before failing completely. There's no time delay but this could be easily added."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests, sys\n",
    "# Following imports are not required\n",
    "import pandas as pd\n",
    "import json\n",
    "\n",
    "\n",
    "# Checks the server response code is correct \n",
    "def checkResponse(resp):\n",
    "    if(resp.status_code == 200):\n",
    "        return True\n",
    "    else:\n",
    "        return False \n",
    "\n",
    "\n",
    "# Carries out the REST get request and returns a combined list of all the results    \n",
    "def jsonRequest(geneName: str, taxId: int, pageSize: int) -> str:\n",
    "        \n",
    "    # Format get URL\n",
    "    geneArg = \"gene=\" + geneName\n",
    "    taxIdArg = \"taxid=\" + taxId\n",
    "    urlFormat = \"format=json\"\n",
    "    startCount = 0\n",
    "    resultsCount = startCount\n",
    "    maxResultsPerPage = pageSize\n",
    "    startAt = \"from=\" + str(startCount)\n",
    "    resultsPerPage = \"size=\" + str(maxResultsPerPage)\n",
    "    entries = []\n",
    "    coordinatesURL = \"https://www.ebi.ac.uk/proteins/api/coordinates\"\n",
    "    fullUrl = coordinatesURL + \"?\" + geneArg + \"&\" + taxIdArg + \"&\" \\\n",
    "    + startAt + \"&\" + resultsPerPage + \"&\" + urlFormat \n",
    "    #print(fullUrl)\n",
    "    responseFail = 0\n",
    "               \n",
    "    while True:\n",
    "        response = requests.get(fullUrl)\n",
    "        if(self.checkResponse(response)):\n",
    "            responseFail = 0\n",
    "            jsonResponse = response.json()\n",
    "            entries.extend(jsonResponse)\n",
    "            resultsCount += maxResultsPerPage\n",
    "                \n",
    "            # Check if we need to do another page\n",
    "            if response.headers[\"x-pagination-totalrecords\"] is None \\\n",
    "            or (resultsCount >= int(response.headers[\"x-pagination-totalrecords\"])):\n",
    "                break\n",
    "            if int(response.headers[\"x-pagination-totalrecords\"]) % pageSize == 0:\n",
    "                print(response.headers)\n",
    "                \n",
    "            # Update page offsets and reformat get URL\n",
    "            startAt = \"from=\" + str(resultsCount)\n",
    "            fullUrl = coordinatesURL + \"?\" + geneArg + \"&\" + taxIdArg + \"&\" \\\n",
    "            + startAt + \"&\" + resultsPerPage + \"&\" + urlFormat \n",
    "                \n",
    "        # Stop requesting from server if have 10 fails\n",
    "        elif(responseFail == 10):\n",
    "            break\n",
    "        # Record a failed response from server\n",
    "        else:\n",
    "            responseFail += 1\n",
    "    print(\"total results count = \" + response.headers[\"x-pagination-totalrecords\"])     \n",
    "    response.close()\n",
    "    return entries\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extracting and Transforming Genomic Coordinates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The next block of code illustrates how to parse the JSON to extract and do a basic transformation of the genomic coordinate data available.\n",
    "\n",
    "Three methods are used but _parseGenomeJson_ does the majority of the work. This method loops through the collection of coordinate elements to extract out specific values such as 'chromosome' from the _genomicLocation_ sub-element.\n",
    "\n",
    "_processGene_ simply finds the primary gene symbol by looping through the gene element.\n",
    "\n",
    "Within _parseGenomeJson_ a call is made to _processExon_ to extract out both the genomic location and equivalent protein location for each exon."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def processGene(gene: []) -> str:\n",
    "    for g in gene:\n",
    "        if(g[\"type\"] == \"primary\"):\n",
    "            return g[\"value\"]  \n",
    "    return \"No_Primary\"\n",
    "    \n",
    "    \n",
    "    \n",
    "def processExon(exon: str) -> str:\n",
    "    exonDat = []\n",
    "    exonDat.append(exon[\"id\"])\n",
    "    exonDat.append(exon[\"genomeLocation\"][\"begin\"][\"position\"])\n",
    "    exonDat.append(exon[\"genomeLocation\"][\"end\"][\"position\"])\n",
    "    exonDat.append(exon[\"proteinLocation\"][\"begin\"][\"position\"])\n",
    "    exonDat.append(exon[\"proteinLocation\"][\"end\"][\"position\"]) \n",
    "    return exonDat\n",
    "    \n",
    "    \n",
    "    \n",
    "      \n",
    "def parseGenomeJson(coordinates: []) -> []: \n",
    "    pData = []\n",
    "\n",
    "    for row in coordinates:\n",
    "        accession = row[\"accession\"]\n",
    "        geneName = processGene(row[\"gene\"])\n",
    "        taxId = row[\"taxid\"]\n",
    "        chromosome = row[\"gnCoordinate\"][0][\"genomicLocation\"][\"chromosome\"]\n",
    "        chrStart = row[\"gnCoordinate\"][0][\"genomicLocation\"][\"start\"]\n",
    "        chrEnd = row[\"gnCoordinate\"][0][\"genomicLocation\"][\"end\"]\n",
    "        ensg = row[\"gnCoordinate\"][0][\"ensemblGeneId\"]\n",
    "        enst = row[\"gnCoordinate\"][0][\"ensemblTranscriptId\"]\n",
    "        ensp = row[\"gnCoordinate\"][0][\"ensemblTranslationId\"]\n",
    "            \n",
    "        if(row[\"gnCoordinate\"][0][\"genomicLocation\"][\"reverseStrand\"]):\n",
    "            strand = \"-\" \n",
    "        else:\n",
    "            strand =\"+\" \n",
    "\n",
    "        exons = row[\"gnCoordinate\"][0][\"genomicLocation\"][\"exon\"]\n",
    "        for e in exons:\n",
    "            processedRow = []\n",
    "            processedRow.append(accession)\n",
    "            processedRow.append(geneName)\n",
    "            processedRow.append(taxId)\n",
    "            processedRow.append(ensg)\n",
    "            processedRow.append(enst)\n",
    "            processedRow.append(ensp)\n",
    "            processedRow.append(chromosome)\n",
    "            processedRow.append(chrStart)\n",
    "            processedRow.append(chrEnd)\n",
    "            processedRow.append(strand)\n",
    "            processedRow.append(len(exons))  \n",
    "            processedRow.extend(processExon(e))\n",
    "            pData.append(processedRow)\n",
    "             \n",
    "    return pData"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Finding and transforming UniProt Sequence Annotations (Features)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We use two methods here to extract and transform UniProt sequence annotations (features).  \n",
    "\n",
    "The second in this code block, _parseFeatureJson_, takes the JSON and loops through the collection of \n",
    "coordinates to extract out the basic data about the protein, the gene and it's genomic coordinates.\n",
    "\n",
    "A second sub-loop takes the feature element extacts general UniProt annotation data related to genomic coordinates for the annotation, like in _parseFeatureJson_, but also calls _processFeature_ to extract the specific genomic coordinates and protein coordinates for the UniProt annotation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def processFeature(feat: str) -> []:\n",
    "        featDat = []\n",
    "        featDat.append(feat[\"genomeLocation\"][\"begin\"][\"position\"])\n",
    "        featDat.append(feat[\"genomeLocation\"][\"end\"][\"position\"])\n",
    "        featDat.append(feat[\"location\"][\"begin\"][\"position\"])\n",
    "        featDat.append(feat[\"location\"][\"end\"][\"position\"])\n",
    "        \n",
    "        evidence = feat[\"evidence\"]\n",
    "        mergedEvid = []\n",
    "        for e in evidence:\n",
    "            if(\"dbReference\" in e):\n",
    "                evidVal = e[\"dbReference\"][\"type\"] + \" \" + e[\"dbReference\"][\"id\"] + \" \" + e[\"code\"]\n",
    "            else:    \n",
    "                evidVal = e[\"code\"] \n",
    "            mergedEvid.append(evidVal)\n",
    "        featDat.append(mergedEvid)\n",
    "        return featDat\n",
    "    \n",
    "\n",
    "def parseFeatureJson(coordinates: [],featureType: str) -> []: \n",
    "        \n",
    "    fData = []\n",
    "    #print(coordinates)\n",
    "    for row in coordinates:\n",
    "        accession = row[\"accession\"]\n",
    "        geneName = self.processGene(row[\"gene\"])\n",
    "        taxId = row[\"taxid\"]\n",
    "        chromosome = row[\"gnCoordinate\"][0][\"genomicLocation\"][\"chromosome\"]\n",
    "        chrStart = row[\"gnCoordinate\"][0][\"genomicLocation\"][\"start\"]\n",
    "        chrEnd = row[\"gnCoordinate\"][0][\"genomicLocation\"][\"end\"]\n",
    "        ensg = row[\"gnCoordinate\"][0][\"ensemblGeneId\"]\n",
    "        enst = row[\"gnCoordinate\"][0][\"ensemblTranscriptId\"]\n",
    "        ensp = row[\"gnCoordinate\"][0][\"ensemblTranslationId\"]\n",
    "            \n",
    "        if(row[\"gnCoordinate\"][0][\"genomicLocation\"][\"reverseStrand\"]):\n",
    "            strand = \"-\" \n",
    "        else:\n",
    "            strand =\"+\" \n",
    "        features = row[\"gnCoordinate\"][0][\"feature\"]\n",
    "        fcnt = 1\n",
    "        for f in features:\n",
    "            # Only want to get the specified featureType (annotation Type)\n",
    "            if(f[\"type\"] != featureType):\n",
    "                continue\n",
    "            processedRow = []\n",
    "            processedRow.append(accession)\n",
    "            processedRow.append(geneName)\n",
    "            processedRow.append(taxId)\n",
    "            processedRow.append(ensg)\n",
    "            processedRow.append(enst)\n",
    "            processedRow.append(ensp)\n",
    "            processedRow.append(chromosome)\n",
    "            processedRow.append(chrStart)\n",
    "            processedRow.append(chrEnd)\n",
    "            processedRow.append(strand)\n",
    "            processedRow.append(featureType)\n",
    "            processedRow.append(\"feature #\" + str(fcnt))\n",
    "            fcnt = fcnt + 1\n",
    "            processedRow.extend(processFeature(f))  \n",
    "            fData.append(processedRow)\n",
    "    return fData "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Main method"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The main method actually has a main class, that isn't defined in this example.  The class could take the main \n",
    "variables as shown but the methods above are called as if not defined in a class.\n",
    "\n",
    "You would have to take all the separate code blocks here and put into a single class to have the python script correctly run (with some minor changes)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == '__main__':\n",
    "    \n",
    "    coordinatesURL = \"https://www.ebi.ac.uk/proteins/api/coordinates\"\n",
    "    geneName = \"fgfr2\"\n",
    "    taxId = \"9606\"\n",
    "    pageSize = 100\n",
    "   \n",
    "    main = ParseCoordinatesJson(coordinatesURL,geneName,taxId,pageSize)\n",
    "    fgfr2Data = jsonRequest(geneName,taxId,pageSize)\n",
    "   \n",
    "    genomeMapping = parseGenomeJson(fgfr2Data)\n",
    "      \n",
    "    for gm in genomeMapping:\n",
    "        print(gm)\n",
    "        \n",
    "    feature = \"disulfide bond\"   \n",
    "    featureMapping = parseFeatureJson(fgfr2Data,feature)   \n",
    "    \n",
    "    for fm in featureMapping:\n",
    "        print(fm) \n",
    "\n",
    "    exit()  "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
